meson.build

```diff
@@ -0,0 +1,463 @@
+  if cpp.has_member('struct AVStream', 'codecpar', prefix : '#include <libavformat/avformat.h>', dependencies : avformat_dep)
+    libgerbera_args += '-DHAVE_AVSTREAM_CODECPAR'
+  endif
+else
+  avcodec_dep = dependency('', required: false)
+  avfilter_dep = dependency('', required: false)
+  avformat_dep = dependency('', required: false)
+  avutil_dep = dependency('', required: false)
+endif
+
+curl_dep = dependency('libcurl', required : get_option('curl'))
+if curl_dep.found()
+  libgerbera_args += '-DHAVE_CURL'
+  libgerbera_args += '-DONLINE_SERVICES'
+  libgerbera_args += '-DATRAILERS'
+  libgerbera_args += '-DSOPCAST'
+endif
+
+exif_dep = dependency('libexif', required : get_option('exif'))
+if exif_dep.found()
+  libgerbera_args += '-DHAVE_LIBEXIF'
+endif
+
+exiv2_dep = dependency('exiv2', required : get_option('exiv2'))
+if exiv2_dep.found()
+  libgerbera_args += '-DHAVE_EXIV2'
+  if exiv2_dep.version().version_compare('<0.28')
+    libgerbera_args += '-D_LIBCPP_ENABLE_CXX17_REMOVED_AUTO_PTR'
+  endif
+endif
+
+ffmpegthumbnailer_dep = dependency('libffmpegthumbnailer', required : get_option('ffmpegthumbnailer'))
+if ffmpegthumbnailer_dep.found()
+  libgerbera_args += '-DHAVE_FFMPEGTHUMBNAILER'
+endif
+
+if get_option('inotify')
+  if host_machine.system() == 'linux'
+    libgerbera_args += '-DHAVE_INOTIFY'
+  else
+    error('inotify is Linux only. Please disable.')
+  endif
+endif
+
+duktape_dep = dependency('duktape', required : get_option('js'))
+if duktape_dep.found()
+  libgerbera_args += '-DHAVE_JS'
+endif
+
+lastfm_dep = dependency('liblastfmlib', required : get_option('lastfm'))
+if lastfm_dep.found()
+  libgerbera_args += '-DHAVE_LASTFMLIB'
+endif
+
+magic_dep = cpp.find_library('magic', required : get_option('magic'))
+if magic_dep.found()
+  libgerbera_args += '-DHAVE_MAGIC'
+endif
+
+if get_option('matroska')
+  ebml_dep = dependency('libebml')
+  matroska_dep = dependency('libmatroska')
+  libgerbera_args += '-DHAVE_MATROSKA'
+else
+  ebml_dep = dependency('', required: false)
+  matroska_dep = dependency('', required: false)
+endif
+
+mysql_dep = dependency('libmariadb', required: get_option('mysql'))
+if mysql_dep.found()
+  libgerbera_args += '-DHAVE_MYSQL'
+endif
+
+if get_option('npupnp')
+  upnp_dep = dependency('libnpupnp')
+  libgerbera_args += '-DUSING_NPUPNP'
+else
+  upnp_dep = dependency('libupnp')
+endif
+
+taglib_dep = dependency('taglib', required : get_option('taglib'))
+if taglib_dep.found()
+  libgerbera_args += '-DHAVE_TAGLIB'
+endif
+
+libgerbera_deps = [
+  avcodec_dep,
+  avfilter_dep,
+  avformat_dep,
+  avutil_dep,
+  curl_dep,
+  duktape_dep,
+  ebml_dep,
+  exif_dep,
+  exiv2_dep,
+  fs_dep,
+  fmt_dep,
+  ffmpegthumbnailer_dep,
+  iconv_dep,
+  lastfm_dep,
+  libatomic_dep,
+  magic_dep,
+  matroska_dep,
+  mysql_dep,
+  nsl_dep,
+  pugixml_dep,
+  socket_dep,
+  spdlog_dep,
+  sqlite3_dep,
+  taglib_dep,
+  upnp_dep,
+  uuid_dep,
+]
+
+compile_info = [ '-DCOMPILE_INFO="' +
```

Comment:
This looks painful and complicated indeed. But this seems to be an array with one element, and later inserted as `'-DCOMPILE_INFO='.join(compile_info)`

I think maybe you meant to not include all those `+ ' ' +`. Just have:

```

  'WITH_MAGIC=' + magic_dep.found().to_string().to_upper(),
  'WITH_MYSQL=' + mysql_dep.found().to_string().to_upper(),
  'WITH_CURL=' + curl_dep.found().to_string().to_upper(),
```

etc.

maybe...

```meson
foreach info: [
    ['MAGIC', magic_dep],
    ['MYSQL', mysql_dep],
    ...
    ]
    found = info[1].found().to_string().to_upper()
    compile_info += 'WITH_@0@=@1@'.format(info[0], found)
endforeach
```

I'm very curious why the current cmake build includes COMPILE_INFO about whether or not the tests were enabled. :/ Likewise about whether the systemd unit was installed...

Given most of these are nearly identical to existing `libgerbera_args += '-DHAVE_*'` it feels like this should be handled in the .cpp source code using preprocessor macros to extend a string. But clearly that is not the current state of the cmake...

Reply:
I tried the first version. I got output like -DCOMPILE_INFO="-DXXX=TRUE-DYY=FALSE-DZZZ=TRUE

Reply:
Ah no

```
-DCOMPILE_INFO=\"WITH_MAGIC=TRUE-DCOMPILE_INFO=WITH_MYSQL=TRUE-DCOMPILE_INFO=WITH_CURL=TRUE-DCOMPILE_INFO=WITH_INOTIFY=TRUE-DCOMPILE_INFO=WITH_JS=TRUE-DCOMPILE_INFO=WITH_TAGLIB=TRUE-DCOMPILE_INFO=WITH_AVCODEC=TRUE-DCOMPILE_INFO=WITH_FFMPEGTHUMBNAILER=TRUE-DCOMPILE_INFO=WITH_EXIF=TRUE-DCOMPILE_I    NFO=WITH_EXIV2=TRUE-DCOMPILE_INFO=WITH_SYSTEMD=TRUE-DCOMPILE_INFO=WITH_LASTFM=TRUE-DCOMPILE_INFO=WITH_DEBUG=TRUE-DCOMPILE_INFO=WITH_TESTS=FALSE\"'
```

Reply:
well, '-DCOMPILE_INFO='.join(compile_info) will get you:
- just compile_info, if there is only one element
- otherwise: compile_info[0] + '-DCOMPILE_INFO=' + compile_info[1] + '-DCOMPILE_INFO=' + compile_info[2] .........


But you have a second copy of -DCOMPILE_INFO= at the beginning of your compile_info string building. So you might as well do compile_info = '-DCOMPILE_INFO=' + ........ followed by add_project_arguments(compile_info, language: 'cpp')

Or, you could do:

```meson
add_project_arguments('-DCOMPILE_INFO=' + ' '.join(compile_info), language : 'cpp')
```

Because the .join() should apply to the single-space separator token.

Reply:
I tried this: 

```
--- a/meson.build
+++ b/meson.build
@@ -203,24 +203,31 @@ libgerbera_deps = [
   uuid_dep,
 ]

-compile_info = [ '-DCOMPILE_INFO="' +
-  'WITH_MAGIC=' + magic_dep.found().to_string().to_upper() + ' ' +
-  'WITH_MYSQL=' + mysql_dep.found().to_string().to_upper() + ' ' +
-  'WITH_CURL=' + curl_dep.found().to_string().to_upper() + ' ' +
-  'WITH_INOTIFY=' + get_option('inotify').to_string().to_upper() + ' ' +
-  'WITH_JS=' + duktape_dep.found().to_string().to_upper() + ' ' +
-  'WITH_TAGLIB=' + taglib_dep.found().to_string().to_upper() + ' ' +
-  'WITH_AVCODEC=' + avcodec_dep.found().to_string().to_upper() + ' ' +
-  'WITH_FFMPEGTHUMBNAILER=' + ffmpegthumbnailer_dep.found().to_string().to_upper() + ' ' +
-  'WITH_EXIF=' + exif_dep.found().to_string().to_upper() +  ' ' +
-  'WITH_EXIV2=' + exiv2_dep.found().to_string().to_upper() + ' ' +
-  'WITH_SYSTEMD=' + get_option('systemd').to_string().to_upper() + ' ' +
-  'WITH_LASTFM=' + lastfm_dep.found().to_string().to_upper() + ' ' +
-  'WITH_DEBUG=' + get_option('debug').to_string().to_upper() + ' ' +
-  'WITH_TESTS=' + get_option('tests').to_string().to_upper() + '"'
-]
-
-add_project_arguments('-DCOMPILE_INFO='.join(compile_info), language : 'cpp')
+compile_info = '-DCOMPILE_INFO="'
+foreach info: [
+  ['WITH_MAGIC', magic_dep.found()],
+  ['WITH_MYSQL', mysql_dep.found()],
+  ['WITH_CURL', curl_dep.found()],
+  ['WITH_INOTIFY', get_option('inotify')],
+  ['WITH_JS', duktape_dep.found()],
+  ['WITH_TAGLIB', taglib_dep.found()],
+  ['WITH_AVCODEC', avcodec_dep.found()],
+  ['WITH_FFMPEGTHUMBNAILER', ffmpegthumbnailer_dep.found()],
+  ['WITH_EXIF', exif_dep.found()],
+  ['WITH_EXIV2', exiv2_dep.found()],
+  ['WITH_SYSTEMD', get_option('systemd')],
+  ['WITH_LASTFM', lastfm_dep.found()],
+  ['WITH_DEBUG', get_option('debug')],
+  ['WITH_TESTS', get_option('tests')],
+  ]
+
+  found = info[1].to_string().to_upper()
+  compile_info += 'WITH_@0@=@1@ '.format(info[0], found)
+endforeach
+
+compile_info += '"'
+
+add_project_arguments(compile_info, language : 'cpp')

 # mangix hacks
 add_project_arguments('-DGIT_COMMIT_HASH=""', language : 'cpp')
```

works. Except the last entry ends in a space.

Reply:
compile_info should be declared as a [] and += adds new array elements.

> Or, you could do:
> ```meson
> add_project_arguments('-DCOMPILE_INFO=' + ' '.join(compile_info), language : 'cpp')
> ```